addons/microsoft_calendar/models/calendar.py

```diff
@@ -347,6 +347,26 @@ def _microsoft_values(self, fields_to_sync, initial_values={}):
 
         return values
 
+    def _ensure_attendees_have_email(self):
+        if self.env.context.get("_attendees_have_email", False):
+            return
+        invalid_event_ids = self.env['calendar.attendee'].search([('event_id', 'in', self.ids), ('partner_id.email', '=', False)]).mapped('event_id')
+        if invalid_event_ids:
+            list_length_limit = 50
+            total_invalid_events = len(invalid_event_ids)
+            invalid_event_ids = invalid_event_ids.sorted(key='start')[:list_length_limit]
+            invalid_events_values = invalid_event_ids.read(['display_time', 'display_name'])
+            invalid_events = ['\t- %s: %s' % (event['display_time'], event['display_name'])
+                              for event in invalid_events_values]
+            invalid_events = '\n'.join(invalid_events)
+            details = "(%d/%d)" % (list_length_limit, total_invalid_events) if list_length_limit < total_invalid_events else "(%d)" % total_invalid_events
```

Comment:
@adwid I think that you can achieve the same result with something like:
```
        invalid_event_ids = self.env['calendar.event'].search(
            domain=[
                ('id', 'in', self.ids),
                ('attendee_ids.partner_id.email', '=', False)],
            fields=['display_time', 'display_name'],
            limit=50,
            order='start')
```
As the id is indexed, this shouldn't be so heavy to make the search, thanks to the ``[('id', 'in', foo)]``. Is the context key really necessary in that case ?

Reply:
@tivisse CPE would like that the warning message mentions the total number of meeting in error (e.g. "Either update the events/attendees or archive these events (50/140):". That way, the user knows that the current list does not contain all incorrect events. Therefore, I need to get all the events. 
However, since `id` is indexed, I think I could still use something based on your suggestion:
```python
        invalid_event_ids = self.env['calendar.event'].search(
            domain=[
                ('id', 'in', self.ids),
                ('attendee_ids.partner_id.email', '=', False)],
            fields=['display_time', 'display_name'])
```
and remove the use of context key.

What do you think ?

Reply:
@tivisse Either I'm missing something, or `fields` is not an available parameter. Moreover, I made a performance test and Arnaud's suggestion seems to be very slightly faster (a difference of ~1ms for 100 events with 5 attendees each). Therefore, I think the current code should not be changed.
So, except if you think that I should still remove the context key, IMHO the PR is ready to be merged.

Reply:
@adwid What I said is that the search must be used instead of a filtered in this case.
I still think that this context key is not needed: it makes the code less clear and as Yannick said, the id search is fast enough.
I would do:
```
invalid_event = self.env['calendar.event'].search([('id', 'in', self.ids), ('attendee_ids.email', '=', False)], order='start').read(['display_time', 'display_name'])
```
because a 1 ms gain is not significant. It is clearer, and it correspond exactly to what you need in your if condition.
You can get rid of the context key. The method can be called several times.

Reply:
@arj-odoo @tivisse Code updated:
- Search on `calendar.event`
- No use of context key

Reply:
Oups, sorry for the delay @adwid 

I meant ``search_read``, not ``search``, to perform the search and the read in a single request.